Add the ability to create JWTs using supertokens

nkshah2 opened the pull request:
## Summary of change
This PR adds a new recipe for creating JWTs using supertokens.
- Adds APIs for creating signed JWTs and for fetching JWKs for verification
- Adds auth0/java-jwt which is used to create JWTs (and verify in unit tests)
- Creates a table for storing keys against signing algorithms

## Related issues
- https://github.com/supertokens/for-zenhub/issues/16

## Related Pull Requests
- https://github.com/supertokens/supertokens-mongodb-plugin/pull/8

## Test Plan
- Unit tests added for JWT creating and JWKs (including verifying using the returned keys)
- API tests added for `/recipe/jwt POST` and `/recipe/jwt/jwks GET`

## Documentation changes
These changes will be made after functionality is implemented in the backend SDK
- [ ] Db schema change
- [ ] API gateway / Hasura integration docs
- [ ] Changing getJWTPayload to getAccessTokenPayload and other similar functions
- [ ] Change the use of the word JWT to access token in the docs (in most places)

## Checklist for important updates
- [x] Changelog has been updated
- [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [x] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] ~~Changes to the version if needed~~
   - ~~In `build.gradle`~~
- [x] Had installed and ran the pre-commit hook
- [x] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [x] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [x] Unit tests for JWKs
- [x] API tests
- [x] Update CHANGELOG
- [x] Add test that tests race condition with duplicate key id and simultaneous writes for nosql plugins
- [x] Add comments to existing isolation tests for storage
- [x] Add logic to create one key for each supported algorithm type on startup (key should only be created if one does not already exist in storage)

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -0,0 +1,124 @@
+        headerClaims.put("kid", keyToUse.keyId);
+
+        long currentTimeInMillis = System.currentTimeMillis();
+        long jwtExpiry = (currentTimeInMillis / 1000) + (jwtValidity); // JWT Expiry is seconds from epoch not millis
```
Since JWTs seem to use seconds for expiry, issued time, not before etc I think it makes sense to use seconds

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWTSigningAPI.java:
```diff
@@ -0,0 +1,76 @@
+        JsonObject payload = InputParser.parseJsonObjectOrThrowError(input, "payload", false);
+        assert payload != null;
+
+        long validity = InputParser.parseLongOrThrowError(input, "validity", false);
```
We can do this but using 0 as validity does not cause any issues. Yes the JWT would not be useable at all if it is 0 but since it isn't a problem should the core validate and throw an error in this case? Its a small change so I don't mind doing it but it doesn't feel necessary.

rishabhpoddar replied:
How about checking if it is >= 0? We don't want to allow negative values?

nkshah2 replied:
Done

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWKAPI.java:
```diff
@@ -0,0 +1,62 @@
+/*
```
So this was probably my OCD kicking in but JWKs is just plural, since the API is to fetch a JWK (even though its an array only one is used) for validation `JWKAPI` seemed like a better name. Im agnostic though so I dont mind either.

nkshah2 replied:
Done

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWKAPI.java:
```diff
@@ -0,0 +1,62 @@
+            reply.add("keys", jwksJsonArray);
+            super.sendJsonResponse(200, reply, resp);
+        } catch (StorageQueryException | StorageTransactionLogicException | NoSuchAlgorithmException
+                | InvalidKeySpecException e) {
```
Both exceptions are thrown when creating a public key from its string representation. For example: 

```java
KeyFactory kf = KeyFactory.getInstance("RSA");
 return (RSAPublicKey) kf.generatePublic(keySpec);
```

would throw `NoSuchAlgorithmException` if the string passed to `KeyFactory.getInstance` is an algorithm that JAVAs cryptography doesnt support (or if it some garbage string for example), and would throw `InvalidKeySpecException` "if the given key specification is inappropriate for this key factory to produce a public key" (picked up from the javadocs for the function)

nkshah2 commented on src/main/java/io/supertokens/Main.java:
```diff
@@ -175,6 +176,7 @@ private void init() throws IOException {
         // init signing keys
         AccessTokenSigningKey.init(this);
         RefreshTokenKey.init(this);
+        JWTSigningKey.init(this);
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+        this.main = main;
+    }
+
+    public static void init(Main main) {
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+        return (JWTSigningKey) main.getResourceDistributor().getResource(RESOURCE_KEY);
+    }
+
+    public synchronized List<JWTSigningKeyInfo> getAllSigningKeys()
```
It doesnt need to be, done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    public synchronized JWTSigningKeyInfo getKeyForAlgorithm(String algorithm)
```
It doesnt need to be, done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    private String getAlgorithmType(String algorithm) throws NoSuchAlgorithmException {
```
Algorithm handling has been moved to an enum `JWTSigningKey.SupportedAlgorithms`, the enum now exposes a way to get algorithm strings and types instead of having functions like this one.

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    public synchronized JWTSigningKeyInfo getKeyForAlgorithm(String algorithm)
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+                   throw new StorageTransactionLogicException(e);
+               }
+
+               JWTSigningKeyInfo keyFromStorage = sqlStorage.getLatestJWTSigningKeyForAlgorithm_Transaction(con, algorithmType);
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+            if (keyFromStorage == null) {
+                long currentTimeInMillis = System.currentTimeMillis();
+                Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+                keyInfo = new JWTSigningKeyInfo(Utils.getUUID(), newKey.publicKey, newKey.privateKey, currentTimeInMillis, algorithm, algorithmType);
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/Start.java:
```diff
@@ -58,6 +60,7 @@
     private static final String APP_ID_KEY_NAME = "app_id";
     private static final String ACCESS_TOKEN_SIGNING_KEY_NAME = "access_token_signing_key";
     private static final String REFRESH_TOKEN_KEY_NAME = "refresh_token_key";
+    private static final String JWT_SIGNING_KEY_NAME = "jwt_signing_token_key";
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
         }
     }
+
+    public static List<JWTSigningKeyInfo> getJWTSigningKeys_Transaction(Start start, Connection con)
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+        }
+    }
+
+    private static class JWTSigningKeyInfoRowMapper implements RowMapper<JWTSigningKeyInfo, ResultSet> {
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+        }
+    }
+
+    public static void setJWTSigningKeyInfo_Transaction(Start start, Connection con, JWTSigningKeyInfo info)
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+        }
+    }
+
+    public static JWTSigningKeyInfo getLatestJWTSigningKey_Transaction(Start start, Connection con)
```
Removed the function

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+        return null;
+    }
+
+    public static JWTSigningKeyInfo getLatestJWTSigningKeyForAlgorithm_Transaction(Start start, Connection con, String algorithm)
```
Removed in favour of `getJWTSigningKeys_Transaction` as discussed

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/Start.java:
```diff
@@ -796,4 +799,37 @@ public long getThirdPartyUsersCount() throws StorageQueryException {
+    }
+
+    @Override
+    public JWTSigningKeyInfo getLatestJWTSigningKeyForAlgorithm_Transaction(TransactionConnection con, String algorithm)
```
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+
+        try (PreparedStatement pst = con.prepareStatement(QUERY)) {
+            pst.setString(1, info.keyId);
+            pst.setString(2, info.publicKey);
```
Done. Introduced two new classes (in the plugin interface) for Symmetric and Asymmetric key handling

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+            pst.setString(3, info.privateKey);
+            pst.setLong(4, info.createdAtTime);
+            pst.setString(5, info.algorithm);
+            pst.setString(6, info.algorithmType);
```
Done, removed algorithmType.

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/JWTSigningQueries.java:
```diff
@@ -0,0 +1,35 @@
+                + "PRIMARY KEY(key_id));";
+    }
+
+    static String getQueryToFetchLatestKey(Start start) {
```
Removed

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -0,0 +1,124 @@
+        return jwks;
+    }
+
+    private static JWTSigningKeyInfo getKeyToUse(Main main, String algorithm)
```
Done, removed this function and used the static functon in JWTSigningKey directly

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -0,0 +1,124 @@
+        jwtPayload.put("exp", jwtExpiry);
+        jwtPayload.put("iat", currentTimeInMillis / 1000); // JWT uses seconds from epoch not millis
+
+        // TODO: Add custom payload
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+
+               sqlStorage.commitTransaction(con);
+               return keyInfo;
+            });
```
Done, since NoSuchAlgorithm is an exception that Java's cryptography throws I am now using a custom UnsupportedAlgorithmException for our use case (the names do seem confusing, maybe I should call it something else?)

nkshah2 replied:
The exception is now called `UnsupportedJWTSigningAlgorithmException`

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWTSigningAPI.java:
```diff
@@ -0,0 +1,76 @@
+        String algorithm = InputParser.parseStringOrThrowError(input, "algorithm", false);
+        assert algorithm != null;
+
+        if (!JWTSigningKey.isJWTAlgorithmSupported(algorithm)) {
```
Done, the validation is now handled when converting the string value to the `JWTSigningKey.SupportAlgorithms` enum when createJWT is called.

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,182 @@
+                       long currentTimeInMillis = System.currentTimeMillis();
+                       Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+                       keyInfo = new JWTSigningKeyInfo(Utils.getUUID(), newKey.publicKey, newKey.privateKey, currentTimeInMillis, algorithm, algorithmType);
+                       sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
```
Done

nkshah2 replied:
Done

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/GeneralQueries.java:
```diff
@@ -392,4 +401,89 @@ public static long getUsersCount(Start start, RECIPE_ID[] includeRecipeIds) thro
+    public static List<JWTSigningKeyInfo> getJWTSigningKeys_Transaction(Start start, Connection con)
+            throws SQLException, StorageQueryException {
+
+        // TODO: does this need locking?
```
Done for in memory DB

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWTSigningAPI.java:
```diff
@@ -0,0 +1,76 @@
+            reply.addProperty("status", "OK");
+            reply.addProperty("jwt", jwt);
+            super.sendJsonResponse(200, reply, resp);
+        } catch (Exception e) {
```
The error is no longer needed, changes to the spec are made in this PR: https://github.com/supertokens/core-driver-interface/pull/17

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/Start.java:
```diff
@@ -796,4 +799,37 @@ public long getThirdPartyUsersCount() throws StorageQueryException {
+        Connection sqlCon = (Connection) con.getConnection();
+        try {
+            GeneralQueries.setJWTSigningKeyInfo_Transaction(this, sqlCon, info);
+        } catch (SQLException e) {
```
Done

rishabhpoddar commented on src/main/java/io/supertokens/inmemorydb/queries/JWTSigningQueries.java:
```diff
@@ -16,20 +16,93 @@
+            String algorithm = result.getString("algorithm");
+
+            if (keyString.contains("|")) {
+                return JWTAsymmetricSigningKeyInfo.withKeyString(keyId, createdAt, algorithm, keyString);
+            } else {
+                return new JWTSymmetricSigningKeyInfo(keyId, createdAt, algorithm, keyString);
```
Why is one using `withKeyString` and the other just calling the constructor?

nkshah2 replied:
For Symmetric keys the key string is the key and does not need any extra computation, for Asymmetric keys its a helper function that splits the keystring with the "|" delimiter and then instantiates using the class constructor

rishabhpoddar replied:
But can't the constructor of the `JWTAsymmetricSigningKeyInfo` always just do that? You can have two constructors in that class - one that accepts just a string, and one that accepts the public and private key (same as the existing constructor)

nkshah2 replied:
I can do that as well, I just prefer having static methods. But for the sake of consistency probably a better idea to go with multiple constructors, ill make the change

nkshah2 replied:
Done (Updated plugin interface PR https://github.com/supertokens/supertokens-plugin-interface/pull/22)

rishabhpoddar commented on src/main/java/io/supertokens/inmemorydb/queries/JWTSigningQueries.java:
```diff
@@ -16,20 +16,93 @@
+            pst.setLong(3, info.createdAtTime);
+            pst.setString(4, info.algorithm);
+            pst.executeUpdate();
+        } catch (SQLException e) {
+            throw new StorageTransactionLogicException(e);
+        }
```
Why are you not throwing the `SQLException` directly?

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+               List<JWTSigningKeyInfo> keys;
+               List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+               if (keysFromStorage == null) {
```
When would `keysFromStorage` be `null`? If there are no keys, shouldn't the function return an empty `List`?

nkshah2 replied:
So my thinking was that from the storage layer we should return a list if there are keys, or null otherwise. The storage layer should not assume that the caller does not want to handle nulls. So instead of the storage layer returning an empty List this method handles null.

rishabhpoddar replied:
I disagree. I think the storage layer should return an empty array. It keeps the types clean (the fewer variations, the better). You should make sure that the storage layer returns an empty list in this case.

nkshah2 replied:
So thinking about it a little more I think you're right, ill make the change

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                        } catch (NoSuchAlgorithmException | DuplicateKeyIdException | UnsupportedJWTSigningAlgorithmException e) {
```
If `DuplicateKeyIdException` is thrown, you need to retry this operation here.

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                return sqlStorage.startTransaction(con -> {
+                    JWTSigningKeyInfo keyInfo = null;
+
+                    List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
```
Have you tried to run a SELECT FOR UPDATE query on the whole table in mysql and postgresql and seen if there is locking being done?

nkshah2 replied:
SELECT FOR UPDATE locks the rows that are indexed (indexed is a wrong word here but for lack of better word, I mean rows that the query would traverse and determine if the row needs to be returned) from a query but doesnt lock the entire table, in theory since SELECT * would "touch" all the rows so it would work. However if the table is empty when the query runs (the first time we fetch keys, in the future if the user deletes all keys etc) then it wouldnt lock anything. Still reading up on this

rishabhpoddar replied:
Your last comment is untrue based on our experimentation and tests :) (Just keeping this comment here in case we read the above in the future..)

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                        // Loop through the keys and find the first one for the algorithm
+                        for (int i = 0; i < keysFromStorage.size(); i++) {
+                            JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                            if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
```
Should you just implement a `.equals` function in the algorithm class so that you can do `currentKey.algorithm.equals(algorithm)`

nkshah2 replied:
Makes sense, but `.equals` implies matching with the same type I feel so using `.equalsString` instead

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                        }
+
+                        // If no key was found create a new one
+                        if (keyInfo == null) {
```
Refactor code to prevent duplication of logic for creation of key.

nkshah2 replied:
Done, moved to independent functions for SQL and NOSQL separately

rishabhpoddar replied:
No i meant not to make it into a function, but to structure the code in a way where you need to run this logic just once per transaction - this is cause we can handle the DuplicateKeyId exception properly, just once. It's not really a big deal as such.. but just easier to read the code in the way I am thinking (I feel)

nkshah2 replied:
Done, im still leaving the logic in the individual method because I think it makes the larger method look a little cleaner but I can change that back if you prefer

rishabhpoddar replied:
I prefer if it's not in a function.

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                    return keyInfo;
+                });
+            } catch (StorageTransactionLogicException e) {
+                if (e.actualException instanceof DuplicateKeyIdException) {
```
`DuplicateKeyIdException` should not be thrown at all.

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                if (keysFromStorage.isEmpty()) {
+                    try {
+                        keyInfo = generateKeyForAlgorithm(algorithm);
+                        boolean success = noSQLStorage.setJWTSigningKeyInfo_Transaction(keyInfo);
```
Wouldn't this also throw `DuplicateKeyIdException`?

nkshah2 replied:
It does, its being handled in the independent function which creates the key now. It loops if that error is thrown and retries

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                    if (keyInfo == null) {
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            boolean success = noSQLStorage.setJWTSigningKeyInfo_Transaction(keyInfo);
```
Refactor code to prevent duplication of logic for creation of key.

nkshah2 replied:
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,247 @@
+                    newKey.privateKey);
+        }
+
+        // Ideally this should never execute because for unsupported algorithms an exception will be thrown before this method
+        // This is just so that we dont need to handle null when we call this method
+        throw new UnsupportedJWTSigningAlgorithmException();
```
This is not good logic. If someone passes null into this function, throw some in built runtime exception and not this specific excption that we have. I believe `UnsupportedJWTSigningAlgorithmException` is thrown when creating the `SupportedAlgorithms` object?

nkshah2 replied:
Throwing IllegalArgument as discussed

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -0,0 +1,164 @@
+            supportedAlgorithm = JWTSigningKey.SupportedAlgorithms.valueOf(algorithm);
+        } catch (IllegalArgumentException e) {
+            // If it enters this block then the string value provided does not match the algorithms we support
+            throw new UnsupportedJWTSigningAlgorithmException();
```
Add a test to make sure it comes here as expected.

nkshah2 replied:
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -56,69 +85,80 @@ public static String createJWTToken(Main main, String algorithm, JsonObject payl
-    }
-
-    private static Algorithm getAlgorithmFromString(Main main, String algorithm, JWTSigningKeyInfo keyToUse) throws NoSuchAlgorithmException, InvalidKeySpecException {
+    private static Algorithm getAuth0Algorithm(JWTSigningKey.SupportedAlgorithms algorithm, JWTSigningKeyInfo keyToUse) throws NoSuchAlgorithmException, InvalidKeySpecException,
```
Its directly using Auth0s package which calls the object "Algorithm" which I thought could be confusing since that is used for the signing algorithm as well, for the lack of a better name I used this one

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/queries/JWTSigningQueries.java:
```diff
@@ -16,20 +16,93 @@
+import java.util.List;
 
 public class JWTSigningQueries {
+    private final static String lockString = "jwt_signing_keys_table_lock";
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningFunctions.java:
```diff
@@ -34,16 +37,42 @@
     public static String createJWTToken(Main main, String algorithm, JsonObject payload, String jwksDomain, long jwtValidity)
             throws StorageQueryException, StorageTransactionLogicException, NoSuchAlgorithmException, InvalidKeySpecException,
-            JWTCreationException {
+            JWTCreationException, UnsupportedJWTSigningAlgorithmException, DuplicateKeyIdException {
```
Done

nkshah2 commented on src/main/java/io/supertokens/webserver/api/jwt/JWTSigningAPI.java:
```diff
@@ -67,10 +66,14 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
+            reply.addProperty("status", "UNSUPPORTED_ALGORITHM_ERROR");
             super.sendJsonResponse(200, reply, resp);
+        } catch (StorageQueryException | StorageTransactionLogicException | NoSuchAlgorithmException
+                | InvalidKeySpecException | DuplicateKeyIdException e) {
```
Done

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -100,83 +112,136 @@ public static JWTSigningKey getInstance(Main main) {
+     * @throws DuplicateKeyIdException If there is an error setting generated keys in storage
+     */
+    public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms algorithm)
+            throws UnsupportedJWTSigningAlgorithmException, StorageQueryException, StorageTransactionLogicException, DuplicateKeyIdException {
```
Done

nkshah2 commented:
# Tests Added
Quick overview of all the cases for which tests are added

### Unit Tests

##### JWT creation
- Testing that normal functioning of jwt creation works fine
- Testing that using an invalid algorithm throws UnsupportedJWTSigningAlgorithmException
- Testing that the JWT header has all the required claims
- Testing that the JWT payload has the required claims
- Testing that user defined payload claims are present in the created JWT

##### Getting JWKs
- Testing that the list of JWKs before JWT creation is empty
- Testing that the list of JWKs after JWT creation is not empty
- Testing that the list of JWKs after JWT creation contains a JWK with the same key id as the `kid` claim in the JWT header
- Testing that all the JWKs in the list can be used to create a valid public key (using modulus and exponent)
- Testing that the public key created from the JWK can be correctly used for JWT verification


### API Tests

##### `/recipe/jwt POST`
- Test that the API throws 400 when the required parameters are missing
- Test that the API returns unsupported algorithm status when an invalid `algorithm` parameter is provided
- Test that the API throws 400 when a negative value for `validity` is provided
- Test that calling the API with a valid request body works fine
- Test that the returned JWT has a valid header
- Test that the returned JWT contains all the user defined payload claims

##### `/recipe/jwt/jwks GET`
- Test that the returned keys list is empty before JWT creation
- Test that the returned keys list is not empty after JWT creation
- Test that the returned list after JWT creation has a key with the same key id as the `kid` claim in the JWT header
- Test that a JWK can be used correctly to verify the returned JWT


# Tests/Changes pending
- [ ] Test to simulate a race condition when writing keys to DB to verify that DuplicateKeyId exception is throws
- [ ] Verify the test for negative validity values. Currently assertion errors are throws as 500 but this error should be handled as a 400

nkshah2 commented on src/main/java/io/supertokens/inmemorydb/Start.java:
```diff
@@ -804,31 +804,24 @@ public long getThirdPartyUsersCount() throws StorageQueryException {
-    public JWTSigningKeyInfo getLatestJWTSigningKeyForAlgorithm_Transaction(TransactionConnection con, String algorithm)
-            throws StorageQueryException {
-        Connection sqlCon = (Connection) con.getConnection();
+            if (e.actualException.getMessage().equals("[SQLITE_CONSTRAINT]  Abort due to constraint violation (UNIQUE constraint failed: "+ Config.getConfig(this).getJWTSigningKeysTable() +".key_id)")) {
```
Done

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,238 @@
+            JWTSigningKeyInfo keyInfo = null;
+
+            while (true) {
+                List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
```
We should discuss how NoSQL logic works. It seems incorrect on a quick glance

nkshah2 replied:
As discussed we will use findOneAndUpdate in the case of mongo for this

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,238 @@
+        while (true) {
+            try {
+                JWTSigningKeyInfo keyInfo = generateKeyForAlgorithm(algorithm);
+                boolean success = storage.setJWTSigningKeyInfo_Transaction(keyInfo);
```
The query for the set should be something like insert keyInfo into .. if there is no entry with the matching algo (we need to do this in an atomica way)

nkshah2 replied:
As discussed we will use findOneAndUpdate in the case of mongo for this

nkshah2 commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -0,0 +1,238 @@
+            List<JWTSigningKeyInfo> keys;
+            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+            if (keysFromStorage == null) {
```
Done

nkshah2 commented:
Status of all tests:
The test that failed was `testThatCoreDriverInterfaceSupportedVersionsAreBeingReturnedByTheAPI` which is expected since the coreDriverInterfaceSupported.json has changed

<img width="630" alt="Screenshot 2021-09-08 at 6 26 31 PM" src="https://user-images.githubusercontent.com/18233774/132513352-a82ca027-8849-4c52-bfaf-3793b7d93410.png">

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -142,7 +131,17 @@ public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms al
+                            try {
+                                keyInfo = generateKeyForAlgorithm(algorithm);
+                                sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                                return keyInfo;
```
No return.. We need to break since we are doing commit below.

rishabhpoddar replied:
Given this bug, i wonder why your tests pass. I think the transaction logic commits on its own if we don't call the commit? Or is it that your tests don't test this case?

nkshah2 replied:
The transaction logic auto commits but fixed it regardless

rishabhpoddar commented on src/main/java/io/supertokens/jwt/JWTSigningKey.java:
```diff
@@ -174,10 +173,21 @@ public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms al
+                                continue;
+                            }
+
+                            return keyInfo;
```
No return.. we need to break

nkshah2 replied:
Done

rishabhpoddar commented:
> The test that failed was testThatCoreDriverInterfaceSupportedVersionsAreBeingReturnedByTheAPI which is expected since the coreDriverInterfaceSupported.json has changed

So fix the test please.

nkshah2 commented:
All current and newly added tests pass

<img width="610" alt="Screenshot 2021-09-08 at 9 02 34 PM" src="https://user-images.githubusercontent.com/18233774/132539637-2516ddee-026e-47f3-9af8-0f447acb21d9.png">

rishabhpoddar commented:
@nkshah2 we also need to add a test to make sure that the nosql transaction logic is working as expected.

nkshah2 commented on src/test/java/io/supertokens/test/jwt/api/JWTSigningAPITest.java:
```diff
@@ -0,0 +1,341 @@
+/*
```
Done

nkshah2 commented on src/test/java/io/supertokens/test/jwt/api/JWTSigningAPITest.java:
```diff
@@ -0,0 +1,341 @@
+        requestBody.addProperty("validity", 3600);
+
+        try {
+            JsonObject response = io.supertokens.httpRequest.HttpRequest
```
Done, also renamed the class to HttpRequestForTesting

rishabhpoddar commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 - New config values `password_reset_token_lifetime`
   and `email_verification_token_lifetime`: https://github.com/supertokens/supertokens-core/issues/297
+- New JWT recipe to create JWT tokens using SuperTokens
```
Update the changelog to specify db schema changes

nkshah2 replied:
Done

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+                 * should throw an error because the _id of the new document already exists in storage
+                 */
+                JWTSigningKeyInfo keyToSet2 = new JWTSymmetricSigningKeyInfo("keyId-1234", 1000, "EC", "somekeystring");
+                noSQLStorage_1.setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(keyToSet2);
```
Add a fail inside the try / catch

nkshah2 replied:
Done

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+
+            JWTRecipeStorage storage = StorageLayer.getJWTRecipeStorage(process.getProcess());
+
+            if (storage.getType() != STORAGE_TYPE.NOSQL_1) {
+                return;
+            }
```
Add comment as to why this is being done.

nkshah2 replied:
Done, added to the test description above the method declaration

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+                                    }
+
+                                    endValueOfCon1.set(keyId + "-1");
+                                } else {
+                                    endValueOfCon1.set(matchingKeyInfo.keyId);
+                                }
```
Remove this if / else.

nkshah2 replied:
Done

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+
+                                break;
+                            } catch (DuplicateKeyIdException e) {
+                                continue;
```
Throw an exception here.

nkshah2 replied:
Done, the test has been refactored. All exceptions that are not expected result in the test failing.

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+
+                                    if (!success) {
+                                        continue;
+                                    }
```
else fail test

nkshah2 replied:
Done

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+
+                                break;
+                            } catch (DuplicateKeyIdException e) {
+                                continue;
```
Throw an error instead of continue

nkshah2 replied:
Done

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+                                continue;
+                            }
+                        }
+                    } catch (Exception ignored) {
+
+                    }
```
Why do we need this try / catch?

nkshah2 replied:
This is because the calls to storage can throw exceptions (StorageQueryException for example), the test has been refactored to result in failure if any unexpected exceptions occur

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+                                continue;
+                            }
+                        }
+                    } catch (Exception ignored) {
+
+                    }
```
Why do we need this try / catch?

nkshah2 replied:
Same as above

rishabhpoddar commented on src/test/java/io/supertokens/test/jwt/storage/JWTKeyStorageTest.java:
```diff
@@ -0,0 +1,271 @@
+     * able to write to storage
+     */
+    @Test
+    public void testThatSettingJWTKeysInParallelWorksAsExpectedForNoSQL() throws Exception {
```
You need to generate two different keys with the same algorithm within each thread right before calling the mongodb set function.

nkshah2 replied:
Done

rishabhpoddar commented:
Generate one key of each algo type on core startup.

nkshah2 commented:
Summary of changes in the last couple commits:
- Added logic to Main.java to generate one key for each supported algorithm on startup
- Make algorithm type and key type in the JWK json forced to all caps (this is because most popular JWT verification libraries seem to expect it this way)
- Modify JWKSTest to account for new startup key generation logic
- Removed `JWKSAPITest2_9.testGettingKeysBeforeCreatingJWTReturnsEmpty` in favour of `JWKSTest.testThatThereAreTheSameNumberOfJWKSAsSupportedAlgorithmsBeforeJWTCreation`

rishabhpoddar commented on src/main/java/io/supertokens/Main.java:
```diff
@@ -176,6 +179,18 @@ private void init() throws IOException {
+            try {
+                JWTSigningKey.getInstance(this).getOrCreateAndGetKeyForAlgorithm(currentAlgorithm);
+            } catch (StorageQueryException | StorageTransactionLogicException e) {
+                // We ignore this because in this case the only time
```
Unfinished comment?

nkshah2 replied:
Done
